Implement owner-only authorization support (optional, enabled via appication.properties)

EricWittmann opened the pull request:
Implemented via CDI interceptor.  Feature can be enabled/disabled at runtime via configuration.  The new property is `registry.auth.owner-only-authorization` and is a boolean.

The authorization is being applied only to the new V2 API.  We need to consider expanding this functionality to the other supported APIs.

famartinrh commented:
looks good, but what about the search or the UI ? I guess the UI may work out of the box if search auth is addressed.

EricWittmann commented:
Search authorization isn't needed - this authorization feature only impacts modifications to artifacts, not fetching them.  The UI is a good point though, certain UI elements should be disabled if this feature is turned on and the user is viewing an artifact they can't modify.  That could probably be a later enhancement though.

I'm more concerned about adding equivalent auth controls to the other APIs.

EricWittmann commented:
OK I have updated the implementation to handle the other APIs.  I also updated the `application.properties` to extend the existing role-based protections to the APIs - that was missing from before.  I verified that I was able to change a global rule using the ccompat API even when authenticating as a user with ONLY `read-only` permissions. 

So now with this PR, standard role-based permissions are being applied to all our supported APIs **and** the new owner-only authorization is being applied to all APIs.

There are some issues with the configuration.  The IBM compat API isn't quite complete (I'm thinking about disabling it by default).  Also there are a couple of POST operations that I think are being over-protected right now (e.g. read-only users should be able to invoke operations that search for an artifact version by POSTing content, but that is currently disalllowed).

We may need to perform a group technical review on this soon.

carlesarnal commented:
LGTM overall. +1 to the technical review on this.

famartinrh commented on app/src/main/java/io/apicurio/registry/auth/AuthorizedInterceptor.java:
```diff
@@ -0,0 +1,143 @@
+    @PostConstruct
+    public void onConstruct() {
+        if (isAuthEnabled()) {
+            log.info("*** Only-only authorization is enabled ***");
```
typo :)

famartinrh commented:
+1 from me, the tests you added are good. We don't have yet integration-tests for auth/authz ... but that should come later

EricWittmann commented:
@famartinrh k8s test failures.  can you have a look?
